core/src/consensus/validator_set/dynamic_validator.rs

```diff
@@ -0,0 +1,195 @@
+        let state = client.state_at_term_begin(parent.into())?;
+        Some(get_validators(&state).unwrap())
+    }
+
+    fn validators(&self, parent: H256) -> Option<Validators> {
+        let client: Arc<ConsensusClient> = self.client.read().as_ref().and_then(Weak::upgrade)?;
+        let block_id = parent.into();
+        if client.current_term_id(block_id)? == 0 {
+            return None
+        }
+        let state = client.state_at(block_id)?;
+        Some(get_validators(&state).unwrap())
+    }
+
+    fn validators_pubkey(&self, parent: H256) -> Option<Vec<Public>> {
+        self.validators(parent).map(|validators| validators.pubkeys())
+    }
+}
+
+impl ValidatorSet for DynamicValidator {
+    fn contains(&self, parent: &H256, public: &Public) -> bool {
+        if let Some(validators) = self.validators_pubkey(*parent) {
+            validators.into_iter().any(|pubkey| pubkey == *public)
+        } else {
+            self.initial_list.contains(parent, public)
+        }
+    }
+
+    fn contains_address(&self, parent: &H256, address: &Address) -> bool {
+        if let Some(validators) = self.validators_pubkey(*parent) {
+            validators.into_iter().any(|pubkey| public_to_address(&pubkey) == *address)
+        } else {
+            self.initial_list.contains_address(parent, address)
+        }
+    }
+
+    fn get(&self, parent: &H256, nonce: usize) -> Public {
+        if let Some(validators) = self.validators_pubkey(*parent) {
+            let n_validators = validators.len();
+            validators.into_iter().nth(nonce % n_validators).unwrap()
+        } else {
+            self.initial_list.get(parent, nonce)
+        }
+    }
+
+    fn get_index(&self, parent: &H256, public: &Public) -> Option<usize> {
+        if let Some(validators) = self.validators_pubkey(*parent) {
+            validators.into_iter().enumerate().find(|(_index, pubkey)| pubkey == public).map(|(index, _)| index)
+        } else {
+            self.initial_list.get_index(parent, public)
+        }
+    }
+
+    fn get_index_by_address(&self, parent: &H256, address: &Address) -> Option<usize> {
+        if let Some(validators) = self.validators_pubkey(*parent) {
+            validators
+                .into_iter()
+                .enumerate()
+                .find(|(_index, pubkey)| public_to_address(pubkey) == *address)
+                .map(|(index, _)| index)
+        } else {
+            self.initial_list.get_index_by_address(parent, address)
+        }
+    }
+
+    fn next_block_proposer(&self, parent: &H256, view: u64) -> Option<Address> {
+        if let Some(validators) = self.validators_pubkey(*parent) {
+            let n_validators = validators.len();
+            let nonce = view as usize % n_validators;
+            Some(public_to_address(validators.get(n_validators - nonce - 1).unwrap()))
+        } else {
+            self.initial_list.next_block_proposer(parent, view)
```

Comment:
The algorithm of `DynamicValidator::next_block_proposer` is different from `ValidatorList::next_block_proposer`. I think `initial_list` is more appropriate to be called `static_round_robin_validators` or `ValidatorList` is renamed to `StaticRoundRobinValidators`. However it is too long.

Reply:
I agree. I need another name that specifies it's behavior, but I cannot think.

Reply:
How about simply RoundRobinValidator?
I think it's easy to guess the validator set is static because the other one is `DynamicValidator`.